Adds Inventory Groups

AlexSCorey opened the pull request:
##### SUMMARY
This PR addresses #5271 #5269 and #5268.  
Adds Inventory Groups Add
Adds Inventory Groups Edit
Adds Inventory Groups Form
Adds Inventory Group Details
Adds api module for Groups.  

I expect that there will likely need to be changes as the Inventory Groups list gets completed.  I expect that there will be some changes to routing, and probably some file structure changes.  

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - UI

##### AWX VERSION
```
9.0.1
```


##### ADDITIONAL INFORMATION
![Screen Shot 2019-12-02 at 10 58 41 AM](https://user-images.githubusercontent.com/39280967/69974487-5b530f80-14f3-11ea-9b18-d0433d6f0e0c.png)
![Screen Shot 2019-12-02 at 10 59 09 AM](https://user-images.githubusercontent.com/39280967/69974488-5b530f80-14f3-11ea-9490-3bfd4ced6923.png)
![Screen Shot 2019-12-02 at 10 58 55 AM](https://user-images.githubusercontent.com/39280967/69974490-5b530f80-14f3-11ea-81f4-41f9ffebaf70.png)

AlexSCorey committed:
Adds Inventory Groups routing --sort of
Adds Inventory Groups Add
Adds Inventory Groups Edit
Adds Inventory Groups Form
Adds api module for Groups
Adds placeholder file for InventoryGroupsList.  This was added to refine routing.  Tgere are no tests for this file yet.

AlexSCorey commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.test.jsx:
```diff
@@ -0,0 +1,84 @@
+    wrapper.find('button[aria-label="Edit"]').simulate('click');
+    expect(wrapper.find('InventoryGroupDetail').length).toBe(1);
+    expect(history.location.pathname).toEqual(
+      '/inventories/inventory/undefined/groups/1/edit'
```
Will update this test to look for the actual id shortly.

mabashian commented:
The label on the group details view for `Variables` should be bold/match the other labels:

<img width="1462" alt="Screen Shot 2019-12-03 at 11 25 36 AM" src="https://user-images.githubusercontent.com/9889020/70069416-bc481980-15bf-11ea-8510-d8f7305cbecb.png">

mabashian commented:
I think we'll want to prompt the user when deleting the group from the details view like we do back on the list view for other resources:

![group_delete](https://user-images.githubusercontent.com/9889020/70069608-0e893a80-15c0-11ea-8306-1f495d7eb2de.gif)

The one click delete can be a little jarring and the prompt protects against errant clicks.

mabashian commented:
Looks like an error is being thrown after saving the edit group form (saving is successful it just fails the redirect).  Expectation is that after save you'd be taken back to the group details view. **Note that the same thing happens if you cancel the form**. 

![group_edit](https://user-images.githubusercontent.com/9889020/70069957-aa1aab00-15c0-11ea-83e1-0e3a3e545f10.gif)

mabashian commented:
I know the mockups show an `x` in the upper right corner of the edit form but we've moved away from that pattern in favor of the `Cancel` button.  We don't need two buttons that do the same thing.  I think the mockups just haven't been updated since we made that decision.

<img width="1679" alt="Screen Shot 2019-12-03 at 11 38 19 AM" src="https://user-images.githubusercontent.com/9889020/70070373-7d1ac800-15c1-11ea-8bf8-8fae2ccb7ff4.png">

It looks like one or two of the other forms also have this button but it should be removed.

AlexSCorey committed:
Addresses PR issues
Adds Delete Modal for deleting from Details view
Adds test for delete modal
Addresses styling for Variables label
Removes X close button from form

marshmalien reviewed (changes requested):
<img width="1116" alt="Screen Shot 2019-12-04 at 2 22 37 PM" src="https://user-images.githubusercontent.com/15881645/70173762-89765200-16a1-11ea-8472-701d6e15d178.png">

* Should the breadcrumb include the group name?
`Inventories / :inventory_name / Groups / :group_name / Details`

* Those tabs look like Inventory tabs. The group view should have 3 tabs:
`Details | Related Groups | Hosts`

* The variables input should be on its own line without spacing to the left and on top. 

* The Group Edit form should navigate the user to Inventory Group Detail page on Save or Cancel.
`Inventories / :inventory_name / Groups / :group_name / Details`

Mockup: https://tower-mockups.testing.ansible.com/patternfly/inventories/inventories-groups-detail/

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupEdit/InventroyGroupEdit.test.jsx:
```diff
@@ -0,0 +1,64 @@
+import { act } from 'react-dom/test-utils';
+import { mountWithContexts, waitForElement } from '@testUtils/enzymeHelpers';
+
+import InventoryGroupEdit from './InventoryGroupEdit';
```
This test file name is spelled incorrectly

marshmalien commented on awx/ui_next/src/screens/Inventory/Inventories.jsx:
```diff
@@ -57,6 +57,13 @@ class Inventories extends Component {
       [`/inventories/inventory/${inventory.id}/groups`]: i18n._(t`Groups`),
+      [`/inventories/inventory/${inventory.id}/groups/add`]: i18n._(
+        t`Create New Group`
+      ),
```
Adding the path and name below will include the group name into the breadcrumb. 
[`/inventories/inventory/${inventory.id}/groups/${group && group.id}`]: `${group && group.name}`,

marshmalien commented on awx/ui_next/src/screens/Inventory/Inventories.jsx:
```diff
@@ -57,6 +57,13 @@ class Inventories extends Component {
+      [`/inventories/inventory/${inventory.id}/groups/${group &&
+        group.id}/details`]: i18n._(t`Details`),
+      [`/inventories/inventory/${inventory.id}/groups/${group &&
+        group.id}/edit`]: `${group && group.name}`,
```
The breadcrumb name for the edit path should be "Edit Details"

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroups.jsx:
```diff
@@ -1,10 +1,42 @@
+function InventoryGroups({ setBreadcrumb, inventory }) {
+  return (
+    <Switch>
+      {[
```
Do we need to wrap the Routes in brackets?

AlexSCorey replied:
no we don't

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,97 @@
+            );
+          }}
+        />,
+        <Route
```
We probably don't need to check for an `inventoryGroup` before showing the "not-found" Route.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupAdd/InventoryGroupAdd.jsx:
```diff
@@ -0,0 +1,36 @@
+    history.push(`/inventories/inventory/${inventory.id}/groups`);
+  };
+  if (error) {
+    return <ContentError />;
```
Form submission errors should be displayed on the Form. It looks like we are showing a < ContentError > and we're passing the error to the InventoryGroupForm. Are we showing a form submission error in 2 places?

AlexSCorey replied:
Yes, that is what is happening.  Consider it fixed.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,146 @@
+    }
+  };
+  if (error) {
+    return <ContentError />;
```
We should show an AlertModal for deletion errors. < ContentError > is reserved for Content Loading errors [(fetching data to initialize a page or populate a list)](https://github.com/ansible/awx/blob/devel/awx/ui_next/CONTRIBUTING.md#handling-api-errors).

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupAdd/InventoryGroupAdd.test.jsx:
```diff
@@ -0,0 +1,60 @@
+  afterEach(() => {
+    wrapper.unmount();
+  });
+  test('InventoryGroupEdit renders successfully', () => {
```
InventoryGroupAdd

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupAdd/InventoryGroupAdd.test.jsx:
```diff
@@ -0,0 +1,60 @@
+    await act(async () => {
+      waitForElement(wrapper, 'isLoading', el => el.length === 0);
+    });
+    expect(history.location.pathname).toEqual('/inventories/1/groups');
```
How is this testing cancel navigation?

AlexSCorey replied:
When the user cancels the form is should send them to `/inventories/inventory/1/group` which would be the inventory groups list.  I do see that the url in the code is wrong based on what I just typed so I'll fix that.

AlexSCorey committed:
Adds Alert Modal, Breadcrumb, Nested Tabs and Refactors PR.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupForm/InventoryGroupForm.test.jsx:
```diff
@@ -0,0 +1,34 @@
+    wrapper.unmount();
+  });
+  test('initially renders successfully', () => {
+    expect(wrapper.length).toBe(1);
```
I think it's good that we check that it mounts once, but the following tests probably don't need to.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroups.jsx:
```diff
@@ -1,10 +1,40 @@
+          />
+        )}
+      />
+      ,
```
we can remove this comma

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,157 @@
+      setError(err);
+    }
+  };
+  if (error) {
```
We probably don't want to remove the CardBody when we show the deleteModal and errorModal. Instead of returning, we can hide/show them below in the CardBody.

AlexSCorey committed:
Testing Improvements and Refactoring

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,134 @@
+
+function InventoryGroups({ i18n, match, setBreadcrumb, inventory, history }) {
+  const [inventoryGroup, setInventoryGroup] = useState(null);
+  const [hasContentLoading, setContentLoading] = useState(true);
```
`setContentLoading` should be renamed to `setHasContentLoading`.

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,134 @@
+    },
+  ];
+  if (hasContentError) {
+    return <ContentError />;
```
You'll want to pass the error  to the `ContentError` component so that it can be displayed : 
```js
return <ContentError error={contentError}/>;
```

marshmalien replied:
Should we pass the error to `< ContentError error={contentError} />` ?

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,134 @@
+function InventoryGroups({ i18n, match, setBreadcrumb, inventory, history }) {
+  const [inventoryGroup, setInventoryGroup] = useState(null);
+  const [hasContentLoading, setContentLoading] = useState(true);
+  const [hasContentError, setHasContentError] = useState(false);
```
This isn't a true/false value. Please change it to `[contentError, setContentError]`.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.test.jsx:
```diff
@@ -0,0 +1,70 @@
+    );
+  });
+  test('expect Return to Groups tab to exist', async () => {
+    await act(async () => {
```
This test is a duplicate

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,134 @@
+    };
+
+    loadData();
+  }, [match.params.groupId, setBreadcrumb, inventory]);
```
It is unclear to me why we need to include `setBreadcrumb` in this list. Shouldn't `match.params.groupId` and `inventory` be enough?

AlexSCorey replied:
The linter throws a warning that says we need to keep it in.  I think we are on the verge of having to re think how we are creating our breadcrumbs because in `/inventories` we have to do a lot of prop drilling for `setBreadcrumb`

jakemcdermott replied:
sounds good to me - let's leave it in.

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,134 @@
+    history.location.pathname.endsWith('edit')
+  ) {
+    cardHeader = null;
+  }
```
There's two layers of conditional assignment for `cardHeader` and most of it is unnecessary. 

In the previous lines, we check `hasContentLoading` and return early if it's true, so you don't need to check it again. The rest can be simplified. Try this instead:

```jsx
let cardHeader = null;
if (
  history.location.pathname.includes('groups/') &&
  !history.location.pathname.endsWith('edit')
) {
  cardHeader = (
    <CardHeader style={{ padding: 0 }}>
      <RoutedTabs history={history} tabsArray={tabsArray} />
      <CardCloseButton
        linkTo={`/inventories/inventory/${inventory.id}/group`}
     />
    </CardHeader>
  );
}
```

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.test.jsx:
```diff
@@ -0,0 +1,70 @@
+    wrapper.unmount();
+  });
+  test('renders successfully', async () => {
+    await act(async () => {
```
does this need to be wrapped in an `act` ?

AlexSCorey replied:
No, removed

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,157 @@
+              <Link
+                to={`/users/${inventoryGroup.summary_fields.modified_by.id}`}
+              >
+                {inventoryGroup.summary_fields.modified_by.username}
```
It is not guaranteed that `modified_by.username` will be defined. You'll want to make sure `modified_by`, `username`, actually exist before referencing them.

jakemcdermott replied:
^ same goes for `created_by`

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupEdit/InventroyGroupEdit.test.jsx:
```diff
@@ -0,0 +1,64 @@
+  });
+  test('cancel should navigate user to Inventory Groups List', async () => {
+    await waitForElement(wrapper, 'isLoading', el => el.length === 0);
+    expect(history.location.pathname).toEqual('/inventories/1/groups');
```
This test doesn't seem to be testing the cancel button. Let's expand it so that the cancel is actually being tested or remove the test.

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.test.jsx:
```diff
@@ -0,0 +1,86 @@
+  beforeEach(async () => {
+    await act(async () => {
+      history = createMemoryHistory({
+        initialEntries: ['/inventories/inventory/1/groups/1/edit'],
```
If we are mounting the InventoryGroupDetail component then I think the history initial entry should be the path of the inventory group detail: `/inventories/inventory/1/groups/1/detail`

AlexSCorey committed:
Fixed filename typo

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,157 @@
+  };
+
+  return (
+    <CardBody style={{ paddingTop: '20px' }}>
```
I prefer `css=` for this kind of thing. It comes with our `styled-components` library and lets you write these like you actually would in a stylesheet instead of with js objects. This one could be `css="padding-top: 20px"`

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,157 @@
+            }
+          />
+        )}
+        {modified_by && modified_by.username && (
```
We should handle the case where `modified`/`created` is defined but `modified_by`/`created_by` is not.  This is how we do it for templates: https://github.com/ansible/awx/blob/devel/awx/ui_next/src/screens/Template/JobTemplateDetail/JobTemplateDetail.jsx#L115-L139

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,142 @@
+          path="*"
+          render={() => {
+            return (
+              !hasContentLoading && (
```
We check `hasContentLoading` above and return early if it is true so it doesn't need to be checked again.

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupsList.jsx:
```diff
@@ -0,0 +1,248 @@
+import InventoryGroupsDeleteModal from '../shared/InventoryGroupsDeleteModal';
+
+const QS_CONFIG = getQSConfig('host', {
+  page: 1,
```
Should this be `getQSConfig('group', {..` ?

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupsList.jsx:
```diff
@@ -0,0 +1,248 @@
+import InventoryGroupItem from './InventoryGroupItem';
+import InventoryGroupsDeleteModal from '../shared/InventoryGroupsDeleteModal';
+
+const QS_CONFIG = getQSConfig('host', {
```
This should be `getQSConfig('group', {`

AlexSCorey committed:
Checks for modified and create in Inv Group Deets
Also includes refactoring for css over style prop and removed some
unnecessary loading checks

marshmalien commented on awx/ui_next/src/screens/Inventory/InventoryGroups/InventoryGroupDetail/InventoryGroupDetail.jsx:
```diff
@@ -0,0 +1,165 @@
+        </span>
+      );
+    } else {
+      createdBy = i18n._(t`${formatDateString(inventoryGroup.created)} by`);
```
What is the value of `createdBy` in the else block? `10/10/2009, 1:17:17 PM by`? I think we should remove the string " by". 

This also applies for `modifiedBy` below.

AlexSCorey committed:
Flattens Inventory File Structure Remove

jakemcdermott commented on awx/ui_next/src/screens/Inventory/InventoryGroup/InventoryGroup.jsx:
```diff
@@ -0,0 +1,140 @@
+      <CardHeader style={{ padding: 0 }}>
+        <RoutedTabs history={history} tabsArray={tabsArray} />
+        <CardCloseButton
+          linkTo={`/inventories/inventory/${inventory.id}/group`}
```
This should be `/inventories/inventory/${inventory.id}/groups`

AlexSCorey committed:
Moves inventoryGroupForm into shared directory
Updates InventoryGroups tests
Adds ContentError functionalist to catch a case where a use might navigate to an Inventory
that isn't associated to the shown inventoryGroup.

jakemcdermott reviewed (commented):
This PR is shaping up nicely. I have no more feedback at this time. We'll wait for @marshmalien to have another look at this before approving and merging.

jakemcdermott reviewed (approved):
https://github.com/ansible/awx/pull/5424#pullrequestreview-332039263

AlexSCorey commented:
recheck

jakemcdermott commented:
recheck

AlexSCorey committed:
Fixes failing zuul test

jakemcdermott commented:
recheck

marshmalien reviewed (approved):
![lotr](https://user-images.githubusercontent.com/15881645/70835115-1f109080-1dca-11ea-9bd1-819e853af87c.gif)
@AlexSCorey You did a great job at trailblazing ahead and finding a solution for "nested tabs". Thanks for sticking through the review with us :)
